File: src/main/java/wav/demon/StatCraft/Commands/SC/SCBucketsFilled.java
package wav.demon.StatCraft.Commands.SC;

import com.mysema.query.Tuple;
import com.mysema.query.sql.SQLQuery;
import org.bukkit.command.CommandSender;
import wav.demon.StatCraft.Commands.ResponseBuilder;
import wav.demon.StatCraft.Magic.BucketCode;
import wav.demon.StatCraft.Querydsl.BucketFill;
import wav.demon.StatCraft.Querydsl.QBucketEmpty;
import wav.demon.StatCraft.Querydsl.QBucketFill;
import wav.demon.StatCraft.Querydsl.QPlayers;
import wav.demon.StatCraft.StatCraft;

import java.util.List;

public class SCBucketsFilled extends SCTemplate {

    public SCBucketsFilled(StatCraft plugin) {
        super(plugin);
        this.plugin.getBaseCommand().registerCommand("bucketsfilled", this);
    }

    @Override
    public boolean hasPermission(CommandSender sender, String[] args) {
        return sender.hasPermission("statcraft.user.bucketsfilled");
    }

    @Override
    public String playerStatResponse(String name, List<String> args) {
        try {
            int id = plugin.getDatabaseManager().getPlayerId(name);
            if (id < 0)
                throw new Exception();

            SQLQuery query = plugin.getDatabaseManager().getNewQuery();
            if (query == null)
                return "Sorry, there seems to be an issue connecting to the database right now.";
            QBucketFill f = QBucketFill.bucketFill;
            List<BucketFill> results = query.from(f).where(f.id.eq(id)).list(f);

            int total;
            int water = 0;
            int lava = 0;
            int milk = 0;

            for (BucketFill bucketFill : results) {
                BucketCode code = BucketCode.fromCode(bucketFill.getType());

                if (code == null)
                    continue;

                switch (code) {
                    case WATER:
                        water = bucketFill.getAmount();
                        break;
                    case LAVA:
                        lava = bucketFill.getAmount();
                        break;
                    case MILK:
                        milk = bucketFill.getAmount();
                        break;
                }
            }

            total = water + lava + milk;

            return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Buckets Filled")
                    .addStat("Total", df.format(total))
                    .addStat("Water", df.format(water))
                    .addStat("Lava", df.format(lava))
                    .addStat("Milk", df.format(milk))
                    .toString();
        } catch (Exception e) {
            return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Buckets Filled")
                    .addStat("Total", String.valueOf(0))
                    .addStat("Water", String.valueOf(0))
                    .addStat("Lava", String.valueOf(0))
                    .addStat("Milk", String.valueOf(0))
                    .toString();
        }
    }

    @Override
    public String serverStatListResponse(int num, List<String> args) {
        SQLQuery query = plugin.getDatabaseManager().getNewQuery();
        if (query == null)
            return "Sorry, there seems to be an issue connecting to the database right now.";
        QBucketEmpty e = QBucketEmpty.bucketEmpty;
        QPlayers p = QPlayers.players;
        List<Tuple> result = query
                .from(e)
                .leftJoin(p)
                .on(e.id.eq(p.id))
                .groupBy(p.name)
                .orderBy(e.amount.sum().desc())
                .limit(num)
                .list(p.name, e.amount.sum());

        return topListResponse("Buckets Filled", result);
    }
}

File: src/main/kotlin/com/demonwav/statcraft/commands/sc/SCBucketsFilled.kt
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * https://www.demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.commands.sc

import com.demonwav.statcraft.StatCraft
import com.demonwav.statcraft.commands.ResponseBuilder
import com.demonwav.statcraft.magic.BucketCode
import com.demonwav.statcraft.querydsl.QBucketFill
import com.demonwav.statcraft.querydsl.QPlayers
import org.bukkit.command.CommandSender
import java.sql.Connection

class SCBucketsFilled(plugin: StatCraft) : SCTemplate(plugin) {

    init {
        plugin.baseCommand.registerCommand("bucketsfilled", this)
    }

    override fun hasPermission(sender: CommandSender, args: Array<out String>?) = sender.hasPermission("statcraft.user.bucketsfilled")

    override fun playerStatResponse(name: String, args: List<String>, connection: Connection): String {
        val id = getId(name) ?: return ResponseBuilder.build(plugin) {
            playerName { name }
            statName { "Buckets Filled" }
            stats["Total"] = "0"
            stats["Water"] = "0"
            stats["Lava"] = "0"
            stats["Milk"] = "0"
        }

        val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError

        val f = QBucketFill.bucketFill

        val results = query.from(f).where(f.id.eq(id)).list(f)

        var water = 0
        var lava = 0
        var milk = 0

        for (bucketFill in results) {
            val code = BucketCode.fromCode(bucketFill.type) ?: continue

            when (code) {
                BucketCode.WATER -> water += bucketFill.amount
                BucketCode.LAVA -> lava += bucketFill.amount
                BucketCode.MILK -> milk += bucketFill.amount
            }
        }
        val total = water + lava + milk

        return ResponseBuilder.build(plugin) {
            playerName { name }
            statName { "Buckets Filled" }
            stats["Total"] = df.format(total)
            stats["Water"] = df.format(water)
            stats["Lava"] = df.format(lava)
            stats["Milk"] = df.format(milk)
        }
    }

    override fun serverStatListResponse(num: Long, args: List<String>, connection: Connection): String {
        val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError

        val f = QBucketFill.bucketFill
        val p = QPlayers.players

        val result = query
            .from(f)
            .innerJoin(p)
            .on(f.id.eq(p.id))
            .groupBy(p.name)
            .orderBy(f.amount.sum().desc())
            .limit(num)
            .list(p.name, f.amount.sum())

        return topListResponse("Buckets Filled", result)
    }
}
